src/providers/retryProvider.ts

```diff
@@ -271,6 +300,9 @@ export class RetryProvider extends ethers.providers.StaticJsonRpcProvider {
         return await this._sendAndValidate(provider, method, params);
       } catch (_err: unknown) {
         error = _err;
+        if (method === "eth_call" && this.callReverted(method, error)) {
```

Comment:
Not a strong opinion, but it seems like our way of modeling responses is a little off here and the solution should be just to fix that modeling issue.

Retries are intended to deal with _invalid responses_ (think 500s). A revert is a _valid response_ that happens to generate an application level error (maybe 400 is the right mental model here). _Valid responses_ do not benefit from retries.

Should we be detecting these at the very bottom level (in _trySend), not throwing, and returning special type that indicates a valid response that we convert into an error in the top-level send method?

I think this will avoid us having to do a ton of special

Reply:
It's a good call; we should improve this. Implementing repeated catching and inspection of errors is clunky, and detecting this at the lowest level and returning a meaningful response upwards is more architecturally pure. I also think there are other cases where we'll want custom behaviour - i.e. I think we could probably also tackle the "header not found" error down here, which would ideally help to reduce a fairly significant source of RPC errors (i.e. as we query multiple providers at a blockTag sourced from only a single provider).

I do admit that at this point I'm mostly just interested to get some feedback on how this affects performance at scale in production; the relative ease that we can surgically insert this current band-aid makes me more comfortable in proceeding with a less-than-ideal solution for now, because I'm pretty confident we'll be back here shortly to make further improvements if this change delivers the anticipated performance boost.